### DISCUSSION TITLE
apply the optimistic Update for whatever arg is provided

### DISCUSSION THREAD

**Discussion by samislam** [reduxjs/redux-toolkit#2598] [Q&A]
I have a table where I can delete users as shown in the image bellow:
[![enter image description here][1]][1]

The query I used to fetch this data looks as follows:
```js
useGetManyUsersQuery(queryString.stringify({ role: 'client' })) // --->  role=market
// or just basically useGetManyUsersQuery('role=client')
```

[![enter image description here][2]][2]

And this is my redux-toolkit query API slice:

```js
import mainApi from './index'
import authSlice from './authSlice'

const userSlice = mainApi.injectEndpoints({
  endpoints: (builder) => ({
    getManyUsers: builder.query({
      query: (APIFeatures = '') => ({
        url: `/users?${APIFeatures}`,
        credentials: 'include',
      }),
      transformResponse: (resBody) => resBody.data,
      providesTags: ['users'],
    }),
    deleteOneUserById: builder.mutation({
      query: (userId) => ({
        url: `/users/${userId}`,
        method: 'DELETE',
        credentials: 'include',
      }),
      onQueryStarted(userId, { dispatch, queryFulfilled }) {
        const patchResult = dispatch(
          mainApi.util.updateQueryData('getManyUsers', /./g, (draft) =>
            draft.filter((doc) => doc.id !== userId)
          )
        )
        queryFulfilled.catch(patchResult.undo)
      },
    }),
  }),
})

export const {
  useGetManyUsersQuery,
  useDeleteOneUserByIdMutation,
} = userSlice

export default userSlice

```

I want to see an optimistic Update immediately after I delete one of the clients (users) shown in the table, but how? do I have to pass the **exact** search filter **everytime** I want to emit an action like *delete*? Is there a way I can apply the optimistic update for *whatever* the search filter is used?

You may have noticed above in the code of `userSlice` --> `deleteOneUserById` endpoint --> `onQueryStarted` --> mainApi.util.updateQueryData that I'm writing a regular expression `/./`. Basically here's the point, I want this argument to match *whatever* the argument is provided in the query, I don't care. 

Exactly here:

[![enter image description here][3]][3]

is there any way I can omit this argument or to let it *match anything* ??

  [1]: https://i.stack.imgur.com/7MwQ1.png
  [2]: https://i.stack.imgur.com/VtGg5.png
  [3]: https://i.stack.imgur.com/wlgth.png

**Comment by markerikson** [comment]
Ah... at least looking at `updateQueryData` specifically, no. The second argument is the _exact_ cache key, because that's what RTKQ will use to look up a field like `state.api.queries["getManyUsers('serializedCacheKeyGoesHere')"]`.  RTKQ _has_ to have that actual arg value - you can't regex that part.

**Reply by samislam** [reply]
yes, that's right, .. how can we fix this what do you think?

**Reply by markerikson** [reply]
Well, you've got `getState` available in the second argument to `onQueryStarted`. Assuming you have info in the store on what `role` is being used, you could pull that out and pass `{role: whatever}` as the arg  to `updateQueryData`.

**Reply by samislam** [reply]
But there might be other query parameters as well, like `sort=descending&populate=room&balance[$gte]=100` all of these, like

**Reply by samislam** [reply]
ok I'll find a way for that, anyway,

**Reply by samislam** [reply]
Ok thanks [markerikson](https://github.com/markerikson). can we create a new feature request for this?

**Reply by markerikson** [reply]
Uh... I'm not sure what "new feature" you're asking for here :)

**Reply by samislam** [reply]
Like to allow a way to ignore the cache key and to update the value in the cache directly, to update the state

**Reply by samislam** [reply]
or to set the value of all the cache entries to the optimistic update result

**Reply by markerikson** [reply]
Afraid that's really not feasible, for the reasons I listed above.

You can't just update random pieces of state.  RTKQ has to know _which_ cache entry you're asking to update, and to do that it has to know the cache args so it can serialize them and look up the right field in state.

**Reply by samislam** [reply]
@markerikson , yes thanks to markerikson, I appreciate your attention to the community

**Comment by phryneas** [comment]
You can use `invalidatedBy` to get all the endpoints & arguments providing a specific tag. So just have all of those provide the same tag and then get them that way.

**Reply by samislam** [reply]
Yes that's right, but I want to apply an [optimistic update](https://redux-toolkit.js.org/rtk-query/usage/manual-cache-updates#:~:text=Optimistic%20Updates%E2%80%8B,request%20is%20still%20in%20flight.). 
To see the changes on the table immediately

**Reply by phryneas** [reply]
Yes, that's what `api.util.selectInvalidatedBy` is made for.

**Comment by phryneas** [comment]
And why exactly do you think you could not do that with the information returned from that approach? I added that api for that.

**Reply by samislam** [reply]
Ok, let me complete some refactoring stuff in my code then I'll come back to the caching and rtk stuff in just a few days. And I'll try that out
